src/vcpkg/binarycaching.cpp

```diff
@@ -1997,9 +1997,24 @@ namespace
                 return add_error(msg::format(msgUnknownBinaryProviderType), segments[0].first);
             }
 
+            static const std::map<StringLiteral, Metrics::DefineMetric> metric_names{
+                {"aws", Metrics::DefineMetric::BinaryCachingAws},
+                {"azblob", Metrics::DefineMetric::BinaryCachingAzBlob},
+                {"cos", Metrics::DefineMetric::BinaryCachingCos},
+                {"default", Metrics::DefineMetric::BinaryCachingDefault},
+                {"files", Metrics::DefineMetric::BinaryCachingFiles},
+                {"gcs", Metrics::DefineMetric::BinaryCachingGcs},
+                {"http", Metrics::DefineMetric::BinaryCachingHttp},
+                {"nuget", Metrics::DefineMetric::BinaryCachingNuget},
+            };
+            auto metrics = LockGuardPtr<Metrics>(g_metrics);
             for (const auto& cache_provider : state->binary_cache_providers)
```

Comment:
(problem exists in current code)

1. This is going to be called over and over for each binary caching source
2. This mixes "actions" with "parsing" (yes we do this in other places, but they're bad as well)

At a glance, I think the right fix is to move the metrics calls outside the `handle_segments()` parser and instead perform this as part of `create_binary_providers_from_configs()` transforming the `BinaryConfigParserState` into `std::vector<std::unique_ptr<IBinaryProvider>>`.